services/inc/storage_streams.h

```diff
@@ -0,0 +1,462 @@
+              address_(address),
+              remaining_(length),
+              offset_(offset) {
+    }
+
+    StorageHalInputStream(module_bounds_location_t location, uintptr_t address, size_t length, size_t offset = 0)
+            : storageId_(HAL_STORAGE_ID_INVALID),
+              address_(address),
+              remaining_(length),
+              offset_(offset) {
+        if (location == MODULE_BOUNDS_LOC_INTERNAL_FLASH) {
+            storageId_ = HAL_STORAGE_ID_INTERNAL_FLASH;
+        } else if (location == MODULE_BOUNDS_LOC_EXTERNAL_FLASH) {
+            storageId_ = HAL_STORAGE_ID_EXTERNAL_FLASH;
+        } else {
+            remaining_ = 0;
+        }
+    }
+
+    virtual ~StorageHalInputStream() = default;
+
+    int read(char* data, size_t size) override {
+        CHECK(peek(data, size));
+        return skip(size);
+    }
+
+    int peek(char* data, size_t size) override {
+        if (!remaining_) {
+            return SYSTEM_ERROR_END_OF_STREAM;
+        }
+        CHECK_TRUE(data, SYSTEM_ERROR_INVALID_ARGUMENT);
+        size = std::min(size, remaining_);
+        CHECK(hal_storage_read(storageId_, address_ + offset_, (uint8_t*)data, size));
+        return size;
+    }
+
+    int skip(size_t size) override {
+        //         size, storageId_, address_, remaining_, offset_);
+        if (!remaining_) {
+            return SYSTEM_ERROR_END_OF_STREAM;
+        }
+        size = std::min(size, remaining_);
+        offset_ += size;
+        remaining_ -= size;
+        return size;
+    }
+
+    int seek(size_t offset) override {
+        //         offset, storageId_, address_, remaining_, offset_);
+        CHECK_TRUE(offset <= (offset_ + remaining_), SYSTEM_ERROR_NOT_ENOUGH_DATA);
+        if (offset < offset_) {
+            remaining_ += (offset_ - offset);
+            offset_ = offset;
+        } else {
+            remaining_ -= (offset - offset_);
+            offset_ = offset;
+        }
+        return offset_;
+    }
+
+    int availForRead() override {
+        return remaining_;
+    }
+
+    int waitEvent(unsigned flags, unsigned timeout) override {
+        if (!flags) {
+            return 0;
+        }
+        if (!(flags & InputStream::READABLE)) {
+            return SYSTEM_ERROR_INVALID_ARGUMENT;
+        }
+        if (!remaining_) {
+            return SYSTEM_ERROR_END_OF_STREAM;
+        }
+        return InputStream::READABLE;
+    }
+
+private:
+    hal_storage_id storageId_;
+    uintptr_t address_;
+    size_t remaining_;
+    size_t offset_;
+};
+
+#if HAL_PLATFORM_FILESYSTEM
+class FileInputStream : public InputStream {
+public:
+    FileInputStream(const char* filename, filesystem_instance_t instance, size_t offset = 0)
+            : fs_(nullptr),
+              file_{},
+              remaining_(0),
+              offset_(offset),
+              isOpen_(false) {
+        fs_ = filesystem_get_instance(instance, nullptr);
+        open(filename);
+    }
```

Comment:
Consider only providing a default constructor and exposing the `open()` method so that the possible error when opening the file can be properly handled by the calling code.

Reply:
I will leave this one as-is just for simplicity of not having to carry around `filename` (see `AssetReader` usage as well).